### ISSUE TITLE
Https request reported as http in the httpcontext.request

### ISSUE THREAD

**Issue by bboshoven** [aspnet/IISIntegration#344]
I'd like to start of by saying that I'm not sure if the issue is related to the IISIntegration-repo or asp.net at all, although I think it's most likely. 

### Setup
I have a Windows 2012 R2 Standard server running IIS 8.5 running the latest ASP.NET Core Module running a asp.net core website (targetted: win7-x64 and net462) with a letsencrypt certificate. 

The main function:
```
            var host = new WebHostBuilder()
                .UseKestrel()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .Build();

            host.Run();
```

The middlewares in the startup (Configure method):
```
            app.UseRequestLocalization(new RequestLocalizationOptions { DefaultRequestCulture = new RequestCulture("nl-NL") });

            loggerFactory.AddConsole(Configuration.GetSection("Logging"));
            loggerFactory.AddDebug();

            app.Use((httpContext, next) => 
            {
                lock (@object)
                {
                    var data = new List<string>
                    {
                        $"=============================================================================",
                        $"Scheme: {httpContext.Request.Scheme}",
                        $"QueryString: {httpContext.Request.QueryString}",
                        $"Path: {httpContext.Request.Path}",
                        $"Host: {httpContext.Request.Host}",
                        $"Protocol: {httpContext.Request.Protocol}",
                        $"PathBase: {httpContext.Request.PathBase}",
                        $"IsHttps: {httpContext.Request.IsHttps}"
                    };
                    System.IO.File.AppendAllLines("log.txt", data);
                }
                return next();
            });

                app.UseRewriter(new RewriteOptions().AddRedirectToHttpsPermanent());
                app.UseStaticFiles(new StaticFileOptions
                {
                    OnPrepareResponse = (context) =>
                    {
                        context.Context.Response.Headers.Add("Cache-Control", "public, max-age=604800");
                    }
                });
            app.UseMvcWithDefaultRoute();
```

Relevant package references (for version numbers):
```
    <PackageReference Include="Microsoft.AspNetCore" Version="1.1.1" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc" Version="1.1.2" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.ViewCompilation" Version="1.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Rewrite" Version="1.0.1" />
    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="1.1.1" />
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="1.1.1" />
```

### The problem
So, whenever I make a request to this site through my browser everything works fine. The IIS logs show this: 
```
2017-03-23 21:57:08 <server-ip> GET / - 80 - <client-ip> Mozilla/5.0+(Windows+NT+10.0;+WOW64)+AppleWebKit/537.36+(KHTML,+like+Gecko)+Chrome/57.0.2987.110+Safari/537.36 - 301 0 0 1999
2017-03-23 21:57:09 <server-ip> GET / - 443 - <client-ip> Mozilla/5.0+(Windows+NT+10.0;+WOW64)+AppleWebKit/537.36+(KHTML,+like+Gecko)+Chrome/57.0.2987.110+Safari/537.36 - 200 0 0 984
```
The output from the Use "middleware" shows:
```
=============================================================================
Scheme: http
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: False
=============================================================================
Scheme: https
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: True
```
Up until now everything seems to work correctly. But whenever I run the [PageSpeed Insights](https://developers.google.com/speed/pagespeed/insights/) tool from Google to the same webpage (so: https://myhost/) I get the following outputs:

IIS Logs: 
```
...
2017-03-23 21:59:30 <server-ip> GET / - 443 - <client-ip> Mozilla/5.0+(X11;+Linux+x86_64)+AppleWebKit/537.36+(KHTML,+like+Gecko;+Google+Page+Speed+Insights)+Chrome/27.0.1453+Safari/537.36 - 301 0 0 15
2017-03-23 21:59:30 <server-ip> GET / - 443 - <client-ip> Mozilla/5.0+(Linux;+Android+6.0.1;+Nexus+5X+Build/MMB29P)+AppleWebKit/537.36+(KHTML,+like+Gecko;+Google+Page+Speed+Insights)+Chrome/27.0.1453+Mobile+Safari/537.36 - 301 0 0 46
...
```
So it seems to be requesting https looking at the port. Although the output of the Use is:
```
...
=============================================================================
Scheme: http
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: False
=============================================================================
Scheme: http
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: False
...
```

So, after Pagespeed Insights does its thing, you obviously get the error that it hit the maximum number of redirects. Now the things I've already noticed are:
* From all the site scanning tools I've tried Google's Pagespeed Insights is the only one failing. Even Google's recent "Mobile Friendly Test" works.
* I've tried running my own redirect middleware to https, same result.
* This problem was introduced since RC2 (to my knowledge).

### How to find a solution?
After the debugging I've done so far I ran out of practical ideas to try. But I'm thinking that sometimes the translation from IIS to the Request-object is going wrong, hence I was thinking the IISIntegration-repo was the best choice. However, since it's only Pagespeed Insights having this problem from what I've found so far, the problem might be there.

In other words: is this a problem here and if not, ideas on how to debug this further?

**Comment by Tratcher** [comment]
This is most likely caused by issues with x-forwarded-proto and related headers. See https://github.com/aspnet/Docs/issues/2384.

Can you show the headers for the affected requests?

**Comment by bboshoven** [comment]
Added the headers.

Normal browser request:
```
Scheme: https
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: True
Headers
Cache-Control: max-age=0
Connection: Keep-Alive
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
Accept-Encoding: gzip, deflate, sdch, br
Accept-Language: en-US,en;q=0.8,nl;q=0.6
Cookie: fullcss=true
Host: <host>
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/56.0.2924.87 Safari/537.36
Upgrade-Insecure-Requests: 1
MS-ASPNETCORE-TOKEN: 78ec4f51-7d2e-4150-a295-ec3c3571b70d
X-Original-Proto: http
X-Original-For: 127.0.0.1:51596
```

Google Pagespeed Insights:
```
Scheme: http
QueryString: 
Path: /
Host: <host>
Protocol: HTTP/1.1
PathBase: 
IsHttps: False
Headers
Cache-Control: no-store, no-cache
Connection: Keep-Alive
Accept: text/html,*/*;q=0.8
Accept-Encoding: gzip,deflate,br
Accept-Language: en-US
Host: <host>
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko; Google Page Speed Insights) Chrome/27.0.1453 Safari/537.36
X-Forwarded-For: 92.111.27.14, 66.249.93.71:47927
MS-ASPNETCORE-TOKEN: 78ec4f51-7d2e-4150-a295-ec3c3571b70d
X-Forwarded-Proto: https
```

**Comment by Tratcher** [comment]
There's an extra value in the X-Forwarded-For header. You need to disable RequireHeaderSymmetry https://github.com/aspnet/BasicMiddleware/blob/a641231f663f5ead72ebb84cac7fbd4f012af64d/src/Microsoft.AspNetCore.HttpOverrides/ForwardedHeadersOptions.cs#L74

**Comment by bboshoven** [comment]
You are right, although I needed to add the relevant ForwardedHeaders as well. 

So adding the following code resolves this issue (Startup.cs -> Configure) (fyi for people running into the same problem):

```
            app.UseForwardedHeaders(new ForwardedHeadersOptions
            {
                RequireHeaderSymmetry = false,
                ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto
            });
```

Closing this issue, thanks for the help.

**Comment by Tratcher** [comment]
UseForwardedHeaders is already included by UseIISIntegration. You can configure its options in ConfigureServices.
https://github.com/aspnet/IISIntegration/blob/adcc6e8d7bfd9d9c27920ac04a4e93eaaaf22564/samples/IISSample/Startup.cs#L19

**Comment by Tratcher** [comment]
It's an IList, call Add, Remove, Clear, etc..
